tsdb/engine/tsm1/cache.go

```diff
@@ -180,37 +187,55 @@ func (c *Cache) Keys() []string {
 
 // Values returns a copy of all values, deduped and sorted, for the given key.
 func (c *Cache) Values(key string) Values {
-	values, needSort := func() (Values, bool) {
-		c.mu.RLock()
-		defer c.mu.RUnlock()
-		e := c.store[key]
-		if e == nil {
-			return nil, false
+	c.mu.Lock()
```

Comment:
Perhaps this is what you had in mind -- try to execute the entire function with a read-lock, but if a write-lock is required, oh well, that's what you gotta do.